src/server.rs

```diff
@@ -520,14 +526,27 @@ impl ReplicationBuffer {
             self.message.set_position(previous_pos);
             self.array_len = 0;
             self.arrays_with_data += 1;
+            self.trailing_empty_arrays = 0;
         } else {
+            self.trailing_empty_arrays += 1;
             self.message.set_position(self.array_pos);
             bincode::serialize_into(&mut self.message, &self.array_len)?;
         }
 
         Ok(())
     }
 
+    /// Crops empty arrays at the end.
+    ///
+    /// Should only be called after all arrays have been written, because
+    /// removed array somewhere the middle cannot be detected during deserialization.
+    fn trim_empty_arrays(&mut self) {
+        let used_len = self.message.get_ref().len()
+            - self.trailing_empty_arrays * mem::size_of_val(&self.array_len);
+        self.message.get_mut().truncate(used_len);
```

Comment:
`self.message.get_ref().len()` always contains tick. So `checked_sub()` won't make it robust.
I would keep it as is since the logic is simple and let tests cover the problem.